Add dynamic untracked memory limits for more precise memory tracking

serxa opened the pull request:
The current memory tracker has an error of up to `untracked_memory_limit = 4MB` per thread. Consider a ClickHouse server running with 8GB RAM and a standard `max_server_memory_usage_to_ram_ratio = 0.9` server setting. If there are 1000 threads (and every thread has 800KB on average) OOM can happen before any query is stopped with `MEMORY_LIMIT_EXCEEDED`.

We need total untracked memory to be proportional to current memory usage regardless of the number of threads. To make it fit into 10% RAM we want an error to be less than e.g. 6%. To achieve it we need `untracked_memory_limit` to be updated dynamically and be proportional to current thread memory usage for threads with small memory footprint:
`untracked_memory_limit = clamp(already_tracked / 16, 4KB, 4MB)`.

### Changelog category (leave one):
- Not for changelog

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
This PR was reverted

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

<details>
    <summary>CI Settings</summary>

**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step
- [ ] <!---ci_include_integration--> Allow: Integration Tests
- [ ] <!---ci_include_stateless--> Allow: Stateless tests
- [ ] <!---ci_include_stateful--> Allow: Stateful tests
- [ ] <!---ci_include_unit--> Allow: Unit tests
- [ ] <!---ci_include_performance--> Allow: Performance tests
- [ ] <!---ci_include_aarch64--> Allow: All with aarch64
- [ ] <!---ci_include_asan--> Allow: All with ASAN
- [ ] <!---ci_include_tsan--> Allow: All with TSAN
- [ ] <!---ci_include_analyzer--> Allow: All with Analyzer
- [ ] <!---ci_include_azure --> Allow: All with Azure
- [ ] <!---ci_include_KEYWORD--> Allow: Add your option here
---
- [ ] <!---ci_exclude_fast--> Exclude: Fast test
- [ ] <!---ci_exclude_integration--> Exclude: Integration Tests
- [ ] <!---ci_exclude_stateless--> Exclude: Stateless tests
- [ ] <!---ci_exclude_stateful--> Exclude: Stateful tests
- [ ] <!---ci_exclude_performance--> Exclude: Performance tests
- [ ] <!---ci_exclude_asan--> Exclude: All with ASAN
- [ ] <!---ci_exclude_tsan--> Exclude: All with TSAN
- [ ] <!---ci_exclude_msan--> Exclude: All with MSAN
- [ ] <!---ci_exclude_ubsan--> Exclude: All with UBSAN
- [ ] <!---ci_exclude_coverage--> Exclude: All with Coverage
- [ ] <!---ci_exclude_aarch64--> Exclude: All with Aarch64
---
- [ ] <!---do_not_test--> do not test (only style check)
- [ ] <!---no_merge_commit--> disable merge-commit (no merge from master before tests)
- [ ] <!---no_ci_cache--> disable CI cache (job reuse)
- [ ] <!---batch_0--> allow: batch 1 for multi-batch jobs
- [ ] <!---batch_1--> allow: batch 2
- [ ] <!---batch_2--> allow: batch 3
- [ ] <!---batch_3_4_5--> allow: batch 4, 5 and 6
</details>

serxa committed:
add dynamic untracked memory limits for more precise memory tracking

robot-clickhouse-ci-1 commented:
<!-- automatic status comment for PR #64423 from ClickHouse/ClickHouse:precise-memtracking -->
*This is an automated comment for commit cea82aab5970eeddad04cbabc27407c0c1dc0ff9 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64423/cea82aab5970eeddad04cbabc27407c0c1dc0ff9/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>⏳ pending</td></tr>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>❌ failure</td></tr>
<tbody>
</table>

serxa committed:
fix tests

serxa committed:
fix

serxa committed:
disable precise memory tracking for some tests

robot-clickhouse committed:
Automatic style fix

serxa committed:
turn off dynamic untracked limit memory for not-attached threads (clients and tests)

serxa committed:
Merge branch 'precise-memtracking' of github.com:ClickHouse/ClickHouse into precise-memtracking

serxa committed:
fix syntax error

serxa committed:
disable precise memory tracking to avoid memory_exceeded exception in test

robot-clickhouse committed:
Automatic style fix

serxa committed:
add hysteresis of untracked memory

serxa committed:
Merge branch 'master' into precise-memtracking

serxa committed:
adjust settings history changes

serxa committed:
Merge branch 'master' into precise-memtracking

serxa commented:
I'm not sure what's wrong with the Upgrade check. Does anyone have an idea? I've added new setting to the history and it no does not show up, but check is still failing.

Changed settings are not reflected in settings changes history (see changed_settings.txt) - FAIL
```
   ┌─name─────────────────────────────────────┬─new_value─┬─old_value─┐
1. │ enable_vertical_final                    │ 0         │ 1         │
2. │ output_format_parquet_use_custom_encoder │ 1         │ 0         │
   └──────────────────────────────────────────┴───────────┴───────────┘
   ```

New settings are not reflected in settings changes history (see new_settings.txt) - FAIL
```
    ┌─name──────────────────────────────────────────┐
 1. │ azure_max_blocks_in_multipart_upload          │
 2. │ allow_experimental_join_condition             │
 3. │ prefer_external_sort_block_bytes              │
 4. │ cross_join_min_rows_to_compress               │
 5. │ cross_join_min_bytes_to_compress              │
 6. │ cast_string_to_dynamic_use_inference          │
 7. │ allow_experimental_dynamic_type               │
 8. │ allow_deprecated_error_prone_window_functions │
 9. │ input_format_force_null_for_omitted_fields    │
10. │ input_format_tsv_crlf_end_of_line             │
    └───────────────────────────────────────────────┘
```

serxa committed:
Merge branch 'master' into precise-memtracking

serxa commented:
Okay, merge master and CI tests rerun fixed the problem. Tests `01676_clickhouse_client_autocomplete` and `02456_progress_tty` seem to be flaky. I think this PR is now ready to be merged.

divanik commented on src/Core/SettingsChangesHistory.h:
```diff
@@ -95,6 +95,7 @@ static std::map<ClickHouseVersion, SettingsChangesHistory::SettingsChanges> sett
               {"hdfs_ignore_file_doesnt_exist", false, false, "Allow to return 0 rows when the requested files don't exist instead of throwing an exception in HDFS table engine"},
               {"azure_ignore_file_doesnt_exist", false, false, "Allow to return 0 rows when the requested files don't exist instead of throwing an exception in AzureBlobStorage table engine"},
               {"s3_ignore_file_doesnt_exist", false, false, "Allow to return 0 rows when the requested files don't exist instead of throwing an exception in S3 table engine"},
+              {"min_untracked_memory", 4_MiB, 4_KiB, "A new setting."},
```
(Suggestion) Maybe it is better to add more elaborate description for the setting

divanik commented on src/Common/CurrentMemoryTracker.cpp:
```diff
@@ -57,6 +57,7 @@ AllocationTrace CurrentMemoryTracker::allocImpl(Int64 size, bool throw_if_memory
             {
                 auto res = memory_tracker->allocImpl(will_be, throw_if_memory_exceeded);
                 current_thread->untracked_memory = 0;
+                current_thread->updateUntrackedMemoryLimit(memory_tracker->get());
```
(Suggestion). I understand the strategy of memory enhancing of this PR. As I understand, if we have several hundreds allocations of several KB we are going to flush it a lot of times. If this part is a perfomance-sensitive one, you can replace the current flushing strategy with the exponential one (with exponent 2, for example) goind from min_untracked_memory to max_untracked memory. This will allow us to have log_2(max_untracked_memory/min_untracked_memory) flushes to the main memory tracker until max_untracked_memory is not riched

serxa replied:
It is exponential, but not with factor 2. The factor is 1.0625 (where 6.25% = 1/16). The idea is to keep total untracked memory to be less than 10% of total memory.

divanik commented on tests/queries/0_stateless/01017_uniqCombined_memory_usage.sql:
```diff
@@ -7,7 +7,8 @@
 
-SET use_uncompressed_cache = 0; 
+SET use_uncompressed_cache = 0;
+SET min_untracked_memory = 4194304; -- 4MiB
```
Could we use string '4MiB' here as in 03030?

divanik commented on tests/integration/test_failed_async_inserts/test.py:
```diff
@@ -45,9 +45,7 @@ def test_failed_async_inserts(started_cluster):
-    select_query = (
-        "SELECT value FROM system.events WHERE event == 'FailedAsyncInsertQuery'"
-    )
+    select_query = "SELECT value FROM system.events WHERE event == 'FailedAsyncInsertQuery' SETTINGS min_untracked_memory = 4194304"
```
Could we use string '4MiB' here as in 03030?

divanik commented on tests/integration/test_settings_constraints_distributed/test.py:
```diff
@@ -136,7 +136,9 @@ def test_select_clamps_settings():
     assert (
-        distributed.query(query, settings={"max_memory_usage": 1})
+        distributed.query(
+            query, settings={"max_memory_usage": 1, "min_untracked_memory": 4194304}
```
Could you replace with 4 * 1024 * 1024 or int(2**22)?

serxa committed:
better description for history of a setting changes

serxa committed:
use Mi suffix to make things obvious

serxa committed:
better

serxa committed:
Merge branch 'master' into precise-memtracking

robot-clickhouse committed:
Automatic style fix

serxa committed:
add docs

serxa committed:
typo

Algunenano commented:
This has broken `test_fetch_memory_usage/test.py::test_huge_column` -> https://s3.amazonaws.com/clickhouse-test-reports/0/9ec1306016568a0a89f6ac2b93face43481d04c4/integration_tests__asan__old_analyzer__[6_6].html

Algunenano commented:
It might have also broken test_failed_async_inserts/test.py::test_failed_async_inserts	https://s3.amazonaws.com/clickhouse-test-reports/0/2da27f4deeb4141b154f16bf4420a1018a97b610/integration_tests__release__[2_4].html

Integration tests were not run in the last execution, so it's hard to say
